ENH: spatial.distance: extending `_distance_pybind` with additional metrics

czgdp1807 opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Continues https://github.com/scipy/scipy/pull/14611

#### What does this implement/fix?
<!--Please explain your changes.-->

#### Additional information
<!--Any additional information you think is important.-->

Performance differences with `main`,

```zsh
| Change   | Before [149f9fd3] <main>   | After [874e3e1c] <pybind11_dist>   |   Ratio | Benchmark (Parameter)                                   |
|----------|----------------------------|------------------------------------|---------|---------------------------------------------------------|
| +        | 2.26±0.01μs                | 2.65±0.02μs                        |    1.18 | spatial.Xdist.time_cdist(10, 'cosine')                  |
| +        | 2.11±0.02μs                | 2.38±0.02μs                        |    1.13 | spatial.Xdist.time_pdist(10, 'cosine')                  |
| +        | 2.17±0.04μs                | 2.36±0.02μs                        |    1.09 | spatial.Xdist.time_cdist(10, 'hamming')                 |
| +        | 1.86±0.02μs                | 1.99±0.02μs                        |    1.07 | spatial.Xdist.time_pdist(10, 'hamming')                 |
| -        | 196±2μs                    | 177±3μs                            |    0.9  | spatial.Xdist.time_pdist(100, 'jensenshannon')          |
| -        | 22.5±0.2ms                 | 19.8±0.2ms                         |    0.88 | spatial.Xdist.time_pdist(1000, 'jensenshannon')         |
| -        | 1.26±0.02ms                | 1.11±0.02ms                        |    0.88 | spatial.Xdist.time_pdist(1000, 'seuclidean')            |
| -        | 44.8±0.3ms                 | 38.1±0.3ms                         |    0.85 | spatial.Xdist.time_cdist(1000, 'jensenshannon')         |
| -        | 105±0.4μs                  | 84.9±0.3μs                         |    0.81 | spatial.Xdist.time_cdist(100, 'mahalanobis')            |
| -        | 2.62±0.03ms                | 2.13±0.03ms                        |    0.81 | spatial.Xdist.time_cdist(1000, 'seuclidean')            |
| -        | 423±7μs                    | 323±1μs                            |    0.76 | spatial.Xdist.time_cdist(100, 'jensenshannon')          |
| -        | 20.9±0.1μs                 | 14.8±0.04μs                        |    0.71 | spatial.Xdist.time_cdist(100, 'cosine')                 |
| -        | 1.15±0.03ms                | 817±20μs                           |    0.71 | spatial.Xdist.time_pdist(1000, 'correlation')           |
| -        | 1.14±0.01ms                | 794±20μs                           |    0.7  | spatial.Xdist.time_pdist(1000, 'cosine')                |
| -        | 62.4±1μs                   | 41.8±0.1μs                         |    0.67 | spatial.Xdist.time_pdist(100, 'mahalanobis')            |
| -        | 2.37±0.07ms                | 1.56±0.07ms                        |    0.66 | spatial.Xdist.time_cdist(1000, 'cosine')                |
| -        | 16.8±0.2μs                 | 11.0±0.06μs                        |    0.65 | spatial.Xdist.time_pdist(100, 'correlation')            |
| -        | 2.36±0.05ms                | 1.51±0.06ms                        |    0.64 | spatial.Xdist.time_cdist(1000, 'correlation')           |
| -        | 37.0±0.5μs                 | 21.6±0.09μs                        |    0.58 | spatial.Xdist.time_cdist(100, 'seuclidean')             |
| -        | 23.1±0.1μs                 | 12.5±0.06μs                        |    0.54 | spatial.Xdist.time_pdist(100, 'seuclidean')             |
| -        | 30.9±0.1μs                 | 15.3±0.2μs                         |    0.49 | spatial.Xdist.time_cdist(100, 'correlation')            |
| -        | 6.52±0.09μs                | 2.49±0.06μs                        |    0.38 | spatial.Xdist.time_pdist(10, 'correlation')             |
| -        | 10.5±0.06μs                | 2.74±0.02μs                        |    0.26 | spatial.Xdist.time_cdist(10, 'correlation')             |
| -        | 11.1±0.08μs                | 2.37±0.05μs                        |    0.21 | spatial.Xdist.time_pdist(10, 'seuclidean')              |
| -        | 13.2±0.2μs                 | 2.66±0.03μs                        |    0.2  | spatial.Xdist.time_cdist(10, 'seuclidean')              |
| -        | 22.5±0.2μs                 | 3.64±0.03μs                        |    0.16 | spatial.Xdist.time_cdist(10, 'mahalanobis')             |
| -        | 20.6±0.1μs                 | 2.98±0.1μs                         |    0.14 | spatial.Xdist.time_pdist(10, 'mahalanobis')             |
| -        | 474±2μs                    | 2.87±0.1μs                         |    0.01 | spatial.XdistWeighted.time_pdist(10, 'correlation')     |
| -        | 405±10μs                   | 2.84±0.05μs                        |    0.01 | spatial.XdistWeighted.time_pdist(10, 'cosine')          |
| -        | 1.04±0.01ms                | 3.30±0.03μs                        |    0    | spatial.XdistWeighted.time_cdist(10, 'correlation')     |
| -        | 879±8μs                    | 3.20±0.03μs                        |    0    | spatial.XdistWeighted.time_cdist(10, 'cosine')          |
| -        | 103±0.2ms                  | 24.7±0.5μs                         |    0    | spatial.XdistWeighted.time_cdist(100, 'correlation')    |
| -        | 86.6±0.2ms                 | 24.2±0.4μs                         |    0    | spatial.XdistWeighted.time_cdist(100, 'cosine')         |
| -        | 4.12±0.06ms                | 4.05±0.2μs                         |    0    | spatial.XdistWeighted.time_cdist(20, 'correlation')     |
| -        | 3.49±0.02ms                | 3.97±0.1μs                         |    0    | spatial.XdistWeighted.time_cdist(20, 'cosine')          |
| -        | 51.2±0.06ms                | 14.9±1μs                           |    0    | spatial.XdistWeighted.time_pdist(100, 'correlation')    |
| -        | 43.3±0.2ms                 | 14.6±1μs                           |    0    | spatial.XdistWeighted.time_pdist(100, 'cosine')         |
| -        | 1.99±0.01ms                | 3.40±0.2μs                         |    0    | spatial.XdistWeighted.time_pdist(20, 'correlation')     |
| -        | 1.66±0.01ms                | 3.27±0.2μs                         |    0    | spatial.XdistWeighted.time_pdist(20, 'cosine')          |
```

Performance difference between this branch and [`pybind11_dist_`](https://github.com/czgdp1807/scipy/tree/pybind11_dist_) (based on https://github.com/scipy/scipy/pull/21263#issuecomment-2251367073),

```zsh
| Change   | Before [5ae12deb] <pybind11_dist_>   | After [fa227c73] <pybind11_dist>   |   Ratio | Benchmark (Parameter)                           |
|----------|--------------------------------------|------------------------------------|---------|-------------------------------------------------|
| -        | 1.94±0ms                             | 1.74±0ms                           |    0.89 | spatial.Xdist.time_cdist(1000, 'cosine')        |
| -        | 87.9±0μs                             | 52.7±0μs                           |    0.6  | spatial.Xdist.time_cdist(100, 'cosine')         |
| -        | 91.2±0μs                             | 49.1±0μs                           |    0.54 | spatial.Xdist.time_cdist(10, 'cosine')          |
```

```zsh
| -        | 75.8±0μs                             | 67.5±0μs                           |    0.89 | spatial.XdistWeighted.time_cdist(20, 'cosine')          |
| -        | 106±0μs                              | 78.5±0μs                           |    0.74 | spatial.XdistWeighted.time_cdist(100, 'cosine')         |
| -        | 76.9±0μs                             | 67.0±0μs                           |    0.87 | spatial.XdistWeighted.time_cdist(10, 'cosine')          |
```

Performance difference between variance computation in C++ (`pybind11_dist_2`) and Python (`pybind11_dist`). The performance gains are significant.

```zsh
| Change   | Before [cbabc5a6] <pybind11_dist>   | After [7900364d] <pybind11_dist_2>   |   Ratio | Benchmark (Parameter)                            |
|----------|-------------------------------------|--------------------------------------|---------|--------------------------------------------------|
| -        | 2.50±0ms                            | 2.25±0ms                             |    0.9  | spatial.Xdist.time_cdist(1000, 'seuclidean')    |
| -        | 151±0μs                             | 56.4±0μs                             |    0.37 | spatial.Xdist.time_cdist(100, 'seuclidean')     |
| -        | 148±0μs                             | 37.8±0μs                             |    0.26 | spatial.Xdist.time_cdist(10, 'seuclidean')      |
```

czgdp1807 commented on scipy/spatial/distance.py:
```diff
@@ -1628,6 +1628,28 @@ def sokalsneath(u, v, w=None):
+        return _cdist_callable(
+            XA, XB, metric=metric, out=out, w=w, **kwargs)
+
+    XA = XA.astype(np.float64)
+    XB = XB.astype(np.float64)
```
If I comment these two lines out then the following test fail. Should I increase the tolerances (as the input is `np.float32` so minor increase the tolerances should be fine?)

```zshrc
========================================================== FAILURES ===========================================================
______________________________________ TestCdist.test_cdist_calling_conventions[cosine] _______________________________________
scipy/spatial/tests/test_distance.py:572: in test_cdist_calling_conventions
    self._check_calling_conventions(X1, X2, metric)
        X1         = array([[253.41678 , 133.03748 , 208.3706  , 226.47488 , 173.87625 ,
        290.92435 , 277.08063 , 260.64362 , 295.87...34 , 293.1113  , 128.6125  ,
        283.1046  , 275.94177 , 182.29692 , 294.8665  , 253.00142 ]],
      dtype=float32)
        X2         = array([[209.0965  , 144.3416  , 164.93024 , 167.71597 , 286.10556 ,
        213.18121 , 127.186516, 279.3654  , 244.08...34 , 287.3545  , 129.70238 ,
        291.80734 , 222.30965 , 134.87086 , 106.290665, 281.29544 ]],
      dtype=float32)
        eo_name    = 'random-float32-data'
        metric     = 'cosine'
        self       = <scipy.spatial.tests.test_distance.TestCdist object at 0x137676780>
scipy/spatial/tests/test_distance.py:551: in _check_calling_conventions
    assert_allclose(y1, y2, rtol=eps, verbose=verbose > 2)
        X1         = array([[253.41678 , 133.03748 , 208.3706  , 226.47488 , 173.87625 ,
        290.92435 , 277.08063 , 260.64362 , 295.87...34 , 293.1113  , 128.6125  ,
        283.1046  , 275.94177 , 182.29692 , 294.8665  , 253.00142 ]],
      dtype=float32)
        X2         = array([[209.0965  , 144.3416  , 164.93024 , 167.71597 , 286.10556 ,
        213.18121 , 127.186516, 279.3654  , 244.08...34 , 287.3545  , 129.70238 ,
        291.80734 , 222.30965 , 134.87086 , 106.290665, 281.29544 ]],
      dtype=float32)
        eps        = 1e-07
        kwargs     = {}
        metric     = 'cosine'
        self       = <scipy.spatial.tests.test_distance.TestCdist object at 0x137676780>
        y1         = array([[0.05484979, 0.08171444, 0.05579393, 0.05524987, 0.09750151,
        0.07073686, 0.08786177, 0.04963051, 0.0519..., 0.03029695, 0.07035159, 0.10272532, 0.06459693,
        0.06945175, 0.04259871, 0.06120552, 0.08087305, 0.10174133]])
        y2         = array([[0.05484983, 0.08171439, 0.05579386, 0.05524984, 0.0975015 ,
        0.07073683, 0.08786174, 0.04963045, 0.0519..., 0.03029696, 0.07035156, 0.10272536, 0.06459692,
        0.06945185, 0.04259878, 0.06120565, 0.08087306, 0.10174136]])
        y3         = array([[0.05484983, 0.08171439, 0.05579386, 0.05524984, 0.0975015 ,
        0.07073683, 0.08786174, 0.04963045, 0.0519..., 0.03029696, 0.07035156, 0.10272536, 0.06459692,
        0.06945185, 0.04259878, 0.06120565, 0.08087306, 0.10174136]])
/Users/czgdp1807/mambaforge/envs/scipy-dev/lib/python3.12/contextlib.py:81: in inner
    return func(*args, **kwds)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-07, atol=0
E   
E   Mismatched elements: 354 / 400 (88.5%)
E   Max absolute difference: 1.54077581e-07
E   Max relative difference: 5.8454892e-06
        args       = (<function assert_allclose.<locals>.compare at 0x14b80d1c0>, array([[0.05484979, 0.08171444, 0.05579393, 0.05524987, 0... 0.03029696, 0.07035156, 0.10272536, 0.06459692,
        0.06945185, 0.04259878, 0.06120565, 0.08087306, 0.10174136]]))
        func       = <function assert_array_compare at 0x110981760>
        kwds       = {'equal_nan': True, 'err_msg': '', 'header': 'Not equal to tolerance rtol=1e-07, atol=0', 'verbose': False}
        self       = <contextlib._GeneratorContextManager object at 0x11092a1e0>
_______________________________________ TestCdist.test_cdist_dtype_equivalence[cosine] ________________________________________
scipy/spatial/tests/test_distance.py:610: in test_cdist_dtype_equivalence
    assert_allclose(y1, y2, rtol=eps, verbose=verbose > 2)
        X1         = array([[ True, False,  True, False, False, False,  True, False, False,
         True, False,  True, False, False,  Tru...    [ True, False,  True,  True, False, False,  True,  True, False,
        False, False, False,  True, False,  True]])
        X2         = array([[ True,  True,  True,  True, False,  True, False, False,  True,
         True,  True,  True,  True,  True,  Tru...    [False,  True, False, False,  True, False,  True, False,  True,
         True,  True, False, False, False,  True]])
        eps        = 1e-07
        metric     = 'cosine'
        new_type   = <class 'numpy.float32'>
        self       = <scipy.spatial.tests.test_distance.TestCdist object at 0x1376a1610>
        test       = (array([[False,  True,  True, ..., False,  True,  True],
       [ True,  True,  True, ...,  True,  True,  True],
     ...ue, False, False]]), [<class 'numpy.uint64'>, <class 'numpy.int64'>, <class 'numpy.float32'>, <class 'numpy.float64'>])
        tests      = [(array([[False,  True,  True, ..., False,  True,  True],
       [ True,  True,  True, ...,  True,  True,  True],
    ... 135.6279  , 122.53102 , ..., 150.81123 , 231.02678 ,
        231.16676 ]], dtype=float32), [<class 'numpy.float64'>])]
        y1         = array([[0.41074435, 0.3827866 , 0.42264973, 0.3827866 , 0.45566895,
        0.22848325, 1.        , 0.5669873 , 0.7278..., 0.6619383 , 0.4023857 , 0.42857143, 0.42857143,
        0.53708995, 0.49290745, 0.78178211, 0.62203553, 0.71428571]])
        y2         = array([[0.41074436, 0.38278661, 0.42264974, 0.38278661, 0.45566897,
        0.22848326, 1.        , 0.56698731, 0.7278..., 0.66193829, 0.40238569, 0.4285714 , 0.4285714 ,
        0.53708995, 0.49290744, 0.7817821 , 0.62203552, 0.7142857 ]])
/Users/czgdp1807/mambaforge/envs/scipy-dev/lib/python3.12/contextlib.py:81: in inner
    return func(*args, **kwds)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-07, atol=0
E   
E   Mismatched elements: 6 / 400 (1.5%)
E   Max absolute difference: 4.74261168e-08
E   Max relative difference: 3.31982928e-07
        args       = (<function assert_allclose.<locals>.compare at 0x14b87f4c0>, array([[0.41074435, 0.3827866 , 0.42264973, 0.3827866 , 0... 0.66193829, 0.40238569, 0.4285714 , 0.4285714 ,
        0.53708995, 0.49290744, 0.7817821 , 0.62203552, 0.7142857 ]]))
        func       = <function assert_array_compare at 0x110981760>
        kwds       = {'equal_nan': True, 'err_msg': '', 'header': 'Not equal to tolerance rtol=1e-07, atol=0', 'verbose': False}
        self       = <contextlib._GeneratorContextManager object at 0x11092a1e0>
=================================================== short test summary info ===================================================
FAILED scipy/spatial/tests/test_distance.py::TestCdist::test_cdist_calling_conventions[cosine] - AssertionError: 
FAILED scipy/spatial/tests/test_distance.py::TestCdist::test_cdist_dtype_equivalence[cosine] - AssertionError: 
================================== 2 failed, 1517 passed, 7 skipped, 20 deselected in 8.94s ===================================
```

lucascolley commented:
cc @fancidev for visibility

fancidev commented:
This looks like a very nice addition!

I’m just a little worried about the added complexity to the already complex implementation. Could there be a way to simplify the code while keeping performant?

*If* the input vectors are already normalized (i.e. their length is one), then the cosine distance boils down to is a simple vector inner product, and the “cdist” between two matrices is simply matrix multiplication. This is exactly how scikit-learn [implements cosine distance](https://github.com/scikit-learn/scikit-learn/blob/70fdc843a4b8182d97a3508c1a426acc5e87e980/sklearn/metrics/pairwise.py#L1094) — it normalizes the input matrices and then computes a matrix multiplication.

Is such simplified code performant? I did a quick test and it was twice fast as the current SciPy implementation starting from “moderate sized” input (a few dozen vector and a few dozen elements per vector). I find this not unexpected as matrix multiplication is usually a heavily optimized operation, while the current cosine distance code employs no “instruction level parallelism”.

So I’d suggest to try normalize the input matrices first (possibly from Python code, for convenience), and then call the C++ routine with normalized matrices as input. The existing infrastructure would be adequate to handle this. An added benefit would be that premature numeric overflow could be avoided (if a carefully written `norm` function is called).

What do you think?

czgdp1807 commented:
I think your idea makes sense. Here's the pure Python implementation for unweighted cosine,

```python
XA = XA.astype(np.float64)
XB = XB.astype(np.float64)
x_rownorm_a = np.linalg.norm(XA, axis=1)
y_rownorm_a = np.linalg.norm(XB, axis=1)
XA = XA/np.tile(np.reshape(x_rownorm_a, (x_rownorm_a.shape[0], 1)), XA.shape[1])
XB = XB/np.tile(np.reshape(y_rownorm_a, (y_rownorm_a.shape[0], 1)), XB.shape[1])
cosine = np.matmul(XA, XB.T)
cosine = np.where(np.fabs(cosine) > 1., np.copysign(1, cosine), cosine)

return np.subtract(1., cosine, out=out)
```

fancidev commented:
Sounds cool! FYI you could pass `keepdims=True` in `np.linalg.norm` to make the return value suitable for division in the next step.

Also could use [std::clamp](https://en.cppreference.com/w/cpp/algorithm/clamp) in C++ to constrain the return value.

czgdp1807 commented:
So I do normalisation of `XA` and `XB` at two levels. One at Python level (in `pybind11_dist_`) and one in C++ level (this branch, by unrolling loops by a factor of 4). So, this branch performs much better (by significant margins, see the results I posted in the description of this PR) than `pybind11_dist_`. So I will keep the approach in this PR. IMO, more code is worth having considering the gains we are seeing.

fancidev commented:
Thanks for the comparison. Let me try run a benchmark between your PR and a simplified version and come back with feedback.

fancidev commented:
> Thanks for the comparison. Let me try run a benchmark between your PR and a simplified version and come back with feedback.

I’m happy to confirm that I observed similar performance comparison as you posted in the top post: Normalization performed in C++ is much faster than normalization performed in Python for input scale up to a few hundred vectors times a few elements per vector. For large scale input, the two have the same speed.

I “think” normalization in Python proves slow because `numpy.linalg.norm` is coded in Python and hence rather inefficient in terms of overhead. Plus, it does *not* avoid premature overflow; so there seems no reason to use it in favor of the (straightforward) C++ implementation.

The overall logic of the PR looks good to me.  My main concern is how to achieve the same result with less code duplication. The PR only touches `cdist`; if `pdist` is also handled, more code duplication would result. This seems a bit overwhelming to me.

It appears the main bottleneck is that the existing infrastructure does not provide a way to process the input *after* its dtype is determined. (If the dtype is known in advance, it would not be too difficult to factor the code to avoid the duplication.) Let me think how to enhance the infrastructure to support that.

Meanwhile, let me leave some comments in the code review for discussion.

czgdp1807 commented:
Well we can add an extra argument in the signature of `cdist` in `distance_pybind.cpp`. Something like, `normalise_input`. If `true` then normalise it (inside an `if` condition). Very lite weight and shouldn't affect performance at all. Let me see if that works.

czgdp1807 commented:
I did [this](https://github.com/scipy/scipy/pull/21263#issuecomment-2255522532) for `cdist` and will do it for `pdist`. The code works just fine and without any performance effects.

fancidev commented:
Adding the `normalise_input` and `centralise_input` parameters sounds a good idea! It looks quite lean now.

Maybe they could be factorized into a Preprocessor parameter, in similar fashion to the Distance objects:

struct CosinePreprocessor {
  template <typename T> 
  operator()(x /* StridedView2D<T> */) … 
};

Another thing is that the input arrays, if not copied, should be kept unchanged to avoid surprise to the caller. Therefore the const_casts in the code has to be changed to avoid mutating the input but instead return a copy and mutate the copy.

For that matter, the original C code does not normalize the input but rather computes the norms and divide them in the distance calculation. This likely improves performance for small scale input (by reducing the copy overhead), but it seems not at all straightforward to implement in the existing infrastructure because the Distance objects are called without any “index” parameter (and adding such a parameter would not be “natural”). If you have any good idea to avoid the normalization it would be interesting to see! But the current performance improvement looks already good to me.

The performance drop for tiny-scale input probably comes from the memory-write overhead during normalization. While I don’t care so much (or at all) about the performance of tiny-scale input, an (orthogonal) way to compensate for some of the performance is to unroll the norm calculation. Your code already does the unrolling *within* a row; but if you look closely in the implementation of transform_reduce_2d, it unrolls *across* rows. The reason I think is that

fancidev commented:
Separately, I’m surprised to find that correlation distance does the centering against the _row_ mean instead of the _column_ mean. (The production code already does that and this PR retains that behavior.) Let me try find some reference (for or against) this behavior and open a separate PR if needed.

czgdp1807 commented:
> Therefore the const_casts in the code has to be modified to return a copy instead.

I tried creating a copy in C++ code inside `pdist_impl` but it gives errors in while copying the data from original to the new copy. So, I create a `np.float64` of the inputs at Python level (some assertion tests fail if I don't do so). See, `_distance_pybind_pdist_cosine`, `_distance_pybind_cdist_cosine`, `_distance_pybind_cdist_correlation`, and `_distance_pybind_pdist_correlation` in the diff of this PR.

> Another thing is that the input arrays, if not copied, should be kept unchanged to avoid surprise to the caller. (And it’s possible for the input array to refer to readonly memory in which case writing to it would crash the program.)

There are tests for this in `scipy`. If the original input is modified some tests fail saying that the original input was modified. As I explained above when we call `cdist` and `pdist` functions for `cosine`, `correlation`, I already create a copy for `XA` and `XB`. Therefore no such tests fail. So we are safe here.

For all other metrics the `const_cast` isn't used at all. Its only used for `cosine` and `correlation`.

fancidev commented:
I think it's great to migrate all remaining metrics from C to C++, but I personally find it easier to review if each metric, or each group of related metrics, are migrated in a separate PR.

Separately, if you run the benchmarks locally via `python dev.py bench`, I'd suggest to delete the two occurrences of `--quick` in `dev.py`. The presence of this switch makes the benchmark run only once, which makes the result less reliable. By removing this switch, the benchmarks are run repeatedly and averaged to produce a more reliable result.

czgdp1807 commented:
> I think it's great to migrate all remaining metrics from C to C++, but I personally find it easier to review if each metric, or each group of related metrics, are migrated in a separate PR.

Other than the metrics I am adding here in this PR, all others have been migrated to C++ infrastructure. The ones I am adding here are tricky and need different forms of pre-processing. So a good design of pre-processing needs all the remaining distance metrics (4 in number I think) to be added at the same time. 

To make reviewing easy for you, what I can do is to add changes in each file in separate commits. I will also add comments for each change so that it is for you to understand. Let me know if that works. Creating separate PRs would be tricky for me.

> Separately, if you run the benchmarks locally via python dev.py bench, I'd suggest to delete the two occurrences of --quick in dev.py. The presence of this switch makes the benchmark run only once, which makes the result less reliable. By removing this switch, the benchmarks are run repeatedly and then averaged to produce a more reliable result.

Done locally. You can see the updated description. The performance gains are significant.

czgdp1807 commented:
With C++ computation of inverse covariance matrix we are seeing significant performance gains in `mahalanobis` distance.

czgdp1807 committed:
MISC: Remove --quick from scipy_bench

czgdp1807 committed:
Added the following metrics,

1. NormaliseInput
2. CentraliseInput
3. CosineDistance
4. ScaleInputForJensenshannon
5. JensenshannonDistance

czgdp1807 committed:
Following changes have been made,

1. Added infrastructure to choose pre-processing of inputs for different types distance metrics

2. Added mahalanobis distance with computation inverse covariance matrix

czgdp1807 committed:
Following metrics replaced with _distance_pybind,

1. cosine
2. correlation
3. seuclidean
4. jensenshannon
5. mahalanobis

czgdp1807 commented:
I have added changes in each file in a separate commit for easy reviewing. I will be out of office tomorrow. But once I come back I will leave comments explaining each change in this PR for additional help.

Meanwhile please feel free to suggest any change.

We are seeing excellent performance gains overall.

czgdp1807 commented:
> I think it's great to migrate all remaining metrics from C to C++, but I personally find it easier to review if each metric, or each group of related metrics, are migrated in a separate PR.

My personal opinion is that changes here should go in full inside SciPy's main branch. However, to make reviewing easy for you I can create separate PRs for each metric in my own fork and you can review it there. What do you say? Once all the PRs in my fork are merged then I can create a new PR here in SciPy and we can merge it directly. Let me know if it makes sense.

I have no problems in creating and merging separate PRs directly in SciPy. Just that changes for each metric are linked with others especially the pre-processing part.

fancidev commented:
Thanks for the comments. I think no need to split into separate PRs now that you’ve already included them in one. Btw I’m not a maintainer so I cannot approve your changes; I’m just taking a look out of curiosity and hope to provide some useful comments. Hope you don’t mind! :-)

czgdp1807 commented:
> Hope you don’t mind! :-)

All review comments are welcome by me. I will address them as early as possible. Please let me know of any possible improvements.

rgommers commented:
> Btw I’m not a maintainer so I cannot approve your changes; I’m just taking a look out of curiosity and hope to provide some useful comments. Hope you don’t mind! :-)

@fancidev your reviews here are much appreciated, as is your investigation of other `spatial.distance` topics (and everything else you contribute). I'd encourage you to continue. Also, I sent you an invite for the triage team so you have more permissions on this repo - that was overdue I think.

fancidev commented:
Thanks very much for your trust and support! @rgommers

fancidev commented:
> All review comments are welcome by me. I will address them as early as possible. Please let me know of any possible improvements.

Among the points I raised in the inline reviews, two technical points are worth addressing I think:

1. Avoid using `const_cast` to cast from const to non-const. Although it “works” in the current way the code is written, it could silently break if changes are made in the future. After all, writing to memory obtained from `const_cast` (that casts away constness) leads to undefined behavior in C++ anyway.

2. Wrap new/delete in an RAII object such as std::vector. The current code “works”, but it could silently leak memory if changes are made in the future that throws an exception before delete.

The key question that needs decision, in my opinion, is what sort of code complexity (resulting from micro-optimization) we’re ready to accept in return for code execution speed gain. This is a totally subjective topic so I could only provide my own view to kick off the discussion. (By micro-optimization I mean optimizations that don’t change the big O complexity but reduce the constant factor.)

The main optimizations in the PR are (1) porting certain operations from Python to C++, and (2) unrolling certain loops in C++ to take advantage of instruction level parallelism (ILP).

czgdp1807 committed:
DEV: Use std::vector instead of new

czgdp1807 committed:
DEV: Remove const_cast usage

czgdp1807 committed:
DEV: Remove unnecessary casting to float64

czgdp1807 commented on scipy/spatial/src/distance_pybind.cpp:
```diff
@@ -70,9 +70,28 @@ void validate_weights(const ArrayDescriptor& w, const T* w_data) {
     const intptr_t num_rows = x.shape[0], num_cols = x.shape[1];
 
+    if( centralise_inputs || normalise_inputs ) {
+        T* in_data_ = const_cast<T*>(in_data);
```
This is a review on an outdated commit. This code is no longer present in the latest diff.

czgdp1807 commented on scipy/spatial/src/distance_pybind.cpp:
```diff
@@ -268,7 +361,8 @@ py::array npy_asarray(const py::handle& obj, int flags = 0) {
+                           bool centralise_inputs) {
     auto x = npy_asarray<scalar_t>(x_obj,
                                    NPY_ARRAY_ALIGNED | NPY_ARRAY_NOTSWAPPED);
     auto out = py::cast<py::array_t<scalar_t>>(out_obj);
```
Removed `const_cast` with the help of `NPY_ARRAY_ENSURECOPY` flag. One thing to be noted is this is only applicable for metrics where the copy is needed.

czgdp1807 committed:
DEV: Remove usage new and delete in distance_metrics.h

czgdp1807 committed:
DEV: Remove wrapper functions

czgdp1807 commented on scipy/spatial/distance.py:
```diff
@@ -1628,6 +1628,40 @@ def sokalsneath(u, v, w=None):
+def _distance_pybind_cdist_cosine(XA, XB, *, out=None, **kwargs):
+    w = kwargs.pop('w', None)
+    if w is not None:
+        w = w / w.sum()
```
Addressed. I have removed the wrapper functions.

czgdp1807 commented on scipy/spatial/distance.py:
```diff
@@ -1628,6 +1628,40 @@ def sokalsneath(u, v, w=None):
+        w = w / w.sum()
+
+    XA = XA.astype(np.float64)
+    XB = XB.astype(np.float64)
```
Addressed.

czgdp1807 commented on scipy/spatial/distance.py:
```diff
@@ -1628,6 +1628,40 @@ def sokalsneath(u, v, w=None):
+    if w is not None:
+        w = w / w.sum()
+
+    XA = XA.astype(np.float64)
```
Addressed.

czgdp1807 commented on scipy/spatial/src/distance_metrics.h:
```diff
@@ -169,6 +169,250 @@ void transform_reduce_2d_(
+            x(i, j) = map(x(i, j), rowscales[i]);
+        }
+    }
+}
```
The three new `transform_reduce_2d_` only preprocess inputs (weighted and unweighted). The last one doesn't have a reduce step because it just needs updating of the input by division, subtraction, etc. The first two do have map-reduce step but only accept `x` (no `y`).

czgdp1807 commented:
I have resolved reviews which are addressed by me. There is only one left https://github.com/scipy/scipy/pull/21263#discussion_r1714070803

czgdp1807 commented:
> In my opinion, the key question that needs decision is what sort of additional code complexity we’re ready to accept in return for code execution speed gain.

True. For very large sized inputs benefits of ILP start to vanish. However for small sized inputs they show up (
can be seen in the benchmarks). I think multiple calls with small scale inputs can show performance differences. Say for example, a million calls to the function with each call having small sized inputs. Then performance gains in each call will add up and the overall timings will be improved. So there ILP will definitely help.

czgdp1807 committed:
DEV: Use ForceUnroll to manually unroll loops

czgdp1807 committed:
DEV: Add a utility function for reducing unrolled loop partial results

czgdp1807 committed:
DEV: Use ForceUnroll loop for unrolling loops

czgdp1807 committed:
MISC: Fix linting errors

czgdp1807 commented:
For lesser code duplication, I have used `ForceUnroll` to manually unroll loops. Now this PR is completely ready from my side. @fancidev If you can review and give it a green light then it would be great. Thanks.
